pipit/readers/hpctoolkit_reader.py

```diff
@@ -4,308 +4,1327 @@
+        while self.file.tell() < end_pointer:
+            # Timestamp (nanoseconds since epoch)
+            timestamp = (
+                int.from_bytes(
+                    self.file.read(8), byteorder=self.byte_order, signed=self.signed
                 )
+                - self.min_time_stamp
+            )
+
+            # Sample calling context id (in experiment.xml)
+            # can use this to get name of function from experiement.xml
+            # Procedure tab
+            context_id = int.from_bytes(
+                self.file.read(4), byteorder=self.byte_order, signed=self.signed
+            )
 
-                # Sample calling context id (in experiment.xml)
-                # can use this to get name of function from experiement.xml
-                # Procedure tab
-                calling_context_id = int.from_bytes(
-                    file.read(4), byteorder=byte_order, signed=signed
+            if context_id == last_id:
+                # nothing changed between samples.
+                # means we don't have to do anything
+                continue
+            elif context_id == 0:
+                # process is idling
+                current_node = None
+            else:
+                # at a new non-idle context
+                current_node = self.meta_reader.node_map[context_id]
+
+            # First we want to close all the "enter" events from the last sample
+            # that aren't still running
+            if last_node is not None:
+                if current_node is None:
+                    common_node: Node = None
+                else:
+                    common_node: Node = current_node.get_intersection(last_node)
+
+                # closing each "enter" column until we reach the common_node
+                while last_node != common_node:
+                    curr_ctx_id = self.meta_reader.nid_to_ctx[last_node._pipit_nid]
+                    context_information = (
+                        self.meta_reader.get_information_from_context_id(curr_ctx_id)
+                    )
+
+                    self.data["Name"].append(str(context_information["function"]))
+                    if context_information["loop_type"]:
+                        self.data["Event Type"].append("Loop Leave")
+                    else:
+                        self.data["Event Type"].append("Leave")
+                    self.data["Timestamp (ns)"].append(timestamp)
+                    self.data["Process"].append(hit[1][1])
+                    self.data["Thread"].append(hit[2][1])
+                    self.data["Host"].append(hit[0][1])
+                    self.data["Node"].append(last_node)
+                    self.data["Source File Name"].append(context_information["file"])
+                    self.data["Source File Line Number"].append(
+                        context_information["line"]
+                    )
+                    self.data["Calling Context ID"].append(curr_ctx_id)
```

Comment:
I think getting the other metric information from cct.db should be doable since you have the context ID here. This part of the HPCToolkit on Hatchet might be useful: https://github.com/hatchet/hatchet/pull/508/files#diff-33827f73eebb1a3e89f4d2d3e91f7dd6ecfd4df0a1deb91ba82f4235770bb257R1536